app-tracking-protection/vpn-impl/src/main/java/com/duckduckgo/mobile/android/vpn/integration/NgVpnNetworkStack.kt

```diff
@@ -0,0 +1,268 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.mobile.android.vpn.integration
+
+import android.content.Context
+import android.os.ParcelFileDescriptor
+import android.util.LruCache
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.appbuildconfig.api.AppBuildConfig
+import com.duckduckgo.appbuildconfig.api.isInternalBuild
+import com.duckduckgo.di.scopes.VpnScope
+import com.duckduckgo.mobile.android.vpn.apps.VpnExclusionList
+import com.duckduckgo.mobile.android.vpn.feature.AppTpFeatureConfig
+import com.duckduckgo.mobile.android.vpn.feature.AppTpSetting
+import com.duckduckgo.mobile.android.vpn.model.TrackingApp
+import com.duckduckgo.mobile.android.vpn.model.VpnTracker
+import com.duckduckgo.mobile.android.vpn.network.VpnNetworkStack
+import com.duckduckgo.mobile.android.vpn.processor.requestingapp.AppNameResolver
+import com.duckduckgo.mobile.android.vpn.processor.tcp.tracker.AppTrackerRecorder
+import com.duckduckgo.mobile.android.vpn.service.TrackerBlockingVpnService
+import com.duckduckgo.mobile.android.vpn.store.VpnDatabase
+import com.duckduckgo.mobile.android.vpn.trackers.AppTrackerRepository
+import com.duckduckgo.mobile.android.vpn.trackers.AppTrackerType
+import com.duckduckgo.vpn.network.api.*
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.anvil.annotations.ContributesMultibinding
+import dagger.SingleInstanceIn
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import timber.log.Timber
+import javax.inject.Inject
+
+@ContributesMultibinding(
+    scope = VpnScope::class,
+    boundType = VpnNetworkStack::class
+)
+@ContributesBinding(
+    scope = VpnScope::class,
+    boundType = VpnNetworkCallback::class
+)
+@SingleInstanceIn(VpnScope::class)
+class NgVpnNetworkStack @Inject constructor(
+    private val context: Context,
+    private val appBuildConfig: AppBuildConfig,
+    private val vpnNetwork: VpnNetwork,
+    private val appTrackerRepository: AppTrackerRepository,
+    private val appNameResolver: AppNameResolver,
+    private val appTrackerRecorder: AppTrackerRecorder,
+    private val appTpFeatureConfig: AppTpFeatureConfig,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val vpnNetworkStackVariantManager: VpnNetworkStackVariantManager,
+    vpnDatabase: VpnDatabase,
+) : VpnNetworkStack, VpnNetworkCallback {
+
+    private val packageManager = context.packageManager
+    private val vpnAppTrackerBlockingDao = vpnDatabase.vpnAppTrackerBlockingDao()
+
+    private var tunnelThread: Thread? = null
+    private var jniContext = 0L
+    private val jniLock = Any()
+    private val dnsResourceCache = LruCache<String, String>(1000)
```

Comment:
The reason/logic behind this cache is the following.
* domains are (DNS) resolved
* new networking layer calls `onDnsResolved` where we cache the resource IP address in `dnsResourceCache`
* new networking layer also calls `onSniResolved` where we cache the resource IP address in `dnsResourceCache`
  * `onSniResolved` is only called when users have DNS over TLS enabled (ie. we can't inspect DNS requests as they are encrypted)
 * After `onDnsResolved`, `isDomainBlocked` is called to know whether a domain should be blocked or not
 * Alternatively, `isAddressBlocked` can also be called to know whether an IP address should be blocked or not
  * This is mainly use to support DNS over TLS and uses the `dnsResourceCache` lookup